horovod/spark/common/store.py

```diff
@@ -0,0 +1,338 @@
+                except OSError:
+                    # Race condition from workers on the same host: ignore
+                    pass
+            yield run_path
+
+        return local_run_path
+
+    def sync_fn(self, run_id):
+        def fn(root_path):
+            pass
+        return fn
+
+    @classmethod
+    def filesystem_prefix(cls):
+        return cls.FS_PREFIX
+
+
+class HDFSStore(PrefixStore):
+    FS_PREFIX = 'hdfs://'
+
+    def __init__(self, prefix_path,
+                 host='default', port=0, user=None, kerb_ticket=None,
+                 driver='libhdfs', extra_conf=None, temp_dir=None, *args, **kwargs):
+        super(HDFSStore, self).__init__(prefix_path, *args, **kwargs)
+
+        self._host = host
+        self._port = port
+        self._user = user
+        self._kerb_ticket = kerb_ticket
+        self._driver = driver
+        self._extra_conf = extra_conf
+        self._temp_dir = temp_dir
+        self._hdfs = self._get_filesystem_fn()()
+
+    def get_filesystem(self):
+        return self._hdfs
+
+    def get_local_output_dir_fn(self, run_id):
+        temp_dir = self._temp_dir
+
+        @contextlib.contextmanager
+        def local_run_path():
+            dirpath = tempfile.mkdtemp(dir=temp_dir)
+            try:
+                yield dirpath
+            finally:
+                shutil.rmtree(dirpath)
+
+        return local_run_path
+
+    def sync_fn(self, run_id):
+        class SyncState(object):
+            def __init__(self):
+                self.fs = None
+                self.uploaded = {}
+
+        state = SyncState()
+        get_filesystem = self._get_filesystem_fn()
+        hdfs_root_path = self.get_run_path(run_id)
+
+        def fn(root_path):
+            if state.fs is None:
+                state.fs = get_filesystem()
+
+            hdfs = state.fs
+            uploaded = state.uploaded
+
+            # We need to swap this prefix from the local path with the absolute path, +1 due to
+            # including the trailing slash
+            prefix = len(root_path) + 1
+
+            for local_dir, dirs, files in os.walk(root_path):
+                hdfs_dir = os.path.join(hdfs_root_path, local_dir[prefix:])
+                for file in files:
+                    local_path = os.path.join(local_dir, file)
+                    modified_ts = int(os.path.getmtime(local_path))
+
+                    if local_path in uploaded:
+                        last_modified_ts = uploaded.get(local_path)
+                        if modified_ts <= last_modified_ts:
+                            continue
+
+                    hdfs_path = os.path.join(hdfs_dir, file)
+                    with open(local_path, 'rb') as f:
+                        hdfs.upload(hdfs_path, f)
+                    uploaded[local_path] = modified_ts
+
+        return fn
+
+    def _get_filesystem_fn(self):
+        host = self._host
+        port = self._port
+        user = self._user
```

Comment:
If `user` is empty, we should use the current os user name.

Reply:
This will be handled by Pyarrow.  In order to prevent users misusing the user variable in some way, I changed this to a `self._hdfs_kwargs` variable to encapsulate all these args.